OPHD/UI/Core/RadioButtonGroup.cpp

```diff
@@ -0,0 +1,71 @@
+#include "RadioButtonGroup.h"
+
+#include "../../Cache.h"
+#include "../../Constants.h"
+
+#include <NAS2D/Utility.h>
+#include <NAS2D/Renderer/Renderer.h>
+#include <NAS2D/MathUtils.h>
+
+#include <algorithm>
+
+
+using namespace NAS2D;
+
+
+RadioButtonGroup::RadioButtonGroup(std::vector<ButtonInfo> buttonInfos)
+{
+	mRadioButtons.reserve(buttonInfos.size());
+	for(auto &buttonInfo : buttonInfos)
+	{
+		NAS2D::Vector<int> offset = {0, 13};
+		offset.y = mRadioButtons.size() * offset.y;
+
+		auto &button = mRadioButtons.emplace_back(this, buttonInfo.name, buttonInfo.delegate);
+		button.visible(visible());
+		button.position(mRect.startPoint() + offset);
+	}
+}
+
+void RadioButtonGroup::onMove(NAS2D::Vector<int> displacement)
+{
+	Control::onMove(displacement);
+
+	for (auto &control : mRadioButtons)
+	{
+		control.position(control.position() + displacement);
+	}
+}
+
+void RadioButtonGroup::update()
+{
+	if (!visible()) { return; }
+
+	for (auto &control : mRadioButtons)
+	{
+		control.update();
+		control.hasFocus(true);
+	}
+}
+
+void RadioButtonGroup::clear()
+{
+	if (mIndex != constants::NO_SELECTION)
+	{
+		mRadioButtons[mIndex].checked(false);
+	}
+	mIndex = constants::NO_SELECTION;
+}
+
+void RadioButtonGroup::select(std::size_t index)
+{
+	clear();
+	mIndex = index;
+	mRadioButtons[index].checked(true);
+}
+
+void RadioButtonGroup::select(RadioButtonGroup::RadioButton* button)
+{
+	auto index = button - mRadioButtons.data();
```

Comment:
I would maybe be tempted to use `std::distance(button, mRadioButtons.data())` here. The division inherent in that subtraction is not super obvious, though would be more expected when calling a method such as `distance`. The two are of course completely equivalent in this case. Just that there is a bit more of a high level abstraction when using `std::distance`.

Using `std::distance` might require an additional header include.

Reply:
I suppose I should draw attention to the note on Complexity from the [`std::distance`](https://en.cppreference.com/w/cpp/iterator/distance) documentation: 

> **Complexity**
>
> Linear.
> However, if `InputIt` additionally meets the requirements of `LegacyRandomAccessIterator`, complexity is constant.

The final note about "complexity is constant" is the one that's relevant here. Basically it's a roundabout way of saying the complexity is the same as the subtraction operation. In this case, the pointer meets the definition of `LegacyRandomAccessIterator`.

Reply:
Agreed, I'm all for code that explains itself, and that distance could indeed add some clarity.
I did need to swap the orders of the parameters to get it to work though:

```
void RadioButtonGroup::select(RadioButtonGroup::RadioButton* button)
{
	auto index = std::distance(mRadioButtons.data(), button);
	select(index);
}
```

I also tried to make sure that this wasn't limited to cases wherein the last item was selected by default, so I tried to select the first item (at position 0), this caused a compilation error because in the case of 0, it needs a cast to make sure it knows which overload to pick:

```
    add(rbg, {mProductGrid.size().x + 12, 25});
    rbg.size({50,50});
    rbg.select(static_cast<std::size_t>(0));
```

Reply:
Good catch on the parameter order to `std::distance`.

----

To resolve the overload resolution, you could potentially have used `0u` instead of the `static_cast`. Very interesting that this came up though. I suppose this is one of the reasons they introduced `nullptr` as a keyword, since it's a typed "0", forcing it to be a pointer. Similarly `0u` would be a typed "0", forcing it to be an unsigned integer.

Of course, there is another change that could eliminate this entirely. Change the parameter from a pointer to a reference. I would actually recommend this anyway, since references are non-nullable, and you really don't want someone to pass `nullptr` to that function. It also has the nice side effect of clearing up any ambiguity over what `0` would mean, since `0` isn't a valid reference.

Reply:
Ah yes, that's a very fitting solution, with the added bonus that it's more robust against nullptr values. Fixed.